internal/api/cluster.go

```diff
@@ -18,26 +19,66 @@ func outputJSON(c *Context, w io.Writer, data interface{}) {
 	encoder := json.NewEncoder(w)
 	err := encoder.Encode(data)
 	if err != nil {
-		c.Logger.WithField("error", err).Error("failed to encode result")
+		c.Logger.WithError(err).Error("failed to encode result")
 	}
 }
 
-// handleGetClusters responds to GET /api/clusters.
+// lockCluster synchronizes access to the given cluster across potentially multiple provisioning
+// servers.
+func lockCluster(c *Context, clusterID string) (*model.Cluster, int, func()) {
+	cluster, err := c.Store.GetCluster(clusterID)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to query cluster")
+		return nil, http.StatusInternalServerError, nil
+	}
+	if cluster == nil {
+		return nil, http.StatusNotFound, nil
+	}
+
+	locked, err := c.Store.LockCluster(clusterID)
+	if err != nil {
+		c.Logger.WithError(err).Error("failed to lock cluster")
+		return nil, http.StatusInternalServerError, nil
+	} else if !locked {
+		c.Logger.Error("failed to acquire lock for cluster")
+		return nil, http.StatusConflict, nil
+	}
+
+	unlockOnce := sync.Once{}
+
+	return cluster, 0, func() {
+		unlockOnce.Do(func() {
+			unlocked, err := c.Store.UnlockCluster(cluster.ID, false)
+			if err != nil {
+				c.Logger.WithError(err).Errorf("failed to unlock cluster")
+			} else if unlocked != true {
+				c.Logger.Warn("failed to release lock for cluster")
+			}
+			c.Logger.Infof("unlocked %v", clusterID)
+		})
+	}
+}
+
+// handleGetClusters responds to GET /api/clusters, returning the specified page of clusters.
 func handleGetClusters(c *Context, w http.ResponseWriter, r *http.Request) {
+	var err error
 	pageStr := r.URL.Query().Get("page")
-	page, err := strconv.Atoi(pageStr)
-	if err != nil {
-		c.Logger.WithField("error", err).Error("failed to parse page")
-		w.WriteHeader(http.StatusBadRequest)
-		return
+	page := 0
```

Comment:
Fix an issue when invoking the API via HTTP (by hand) and forgetting to pass the `page` parameter.